### ISSUE TITLE
AWS CNI 1.2.1 throws getsockopt: connection refused

### ISSUE THREAD

**Issue by huynt1979** [aws/amazon-vpc-cni-k8s#236]
I'm following the EKS guide: https://docs.aws.amazon.com/eks/latest/userguide/launch-workers.html in one of our Development enrironment. EKS Worker runs on Ubuntu 16.04 in our private VPC. The AWS CNI version is 1.2.1. I have also verified all the security groups as well as RBAC roles required for the EKS worker setup.

The problem for us is every single pod we try to bring up all ends up dead as they all fail liveness probes with **getsockopt: connection refused**. For example, the **kube-dns** pod in **kube-system** namespace is as below:

```
kubectl -n kube-system describe pods kube-dns
Name:           kube-dns-7cc87d595-mb5v7
Namespace:      kube-system
Node:           ip-10-55-142-134.us-west-2.compute.internal/10.55.142.134
Start Time:     Tue, 20 Nov 2018 16:33:57 -0800
Labels:         eks.amazonaws.com/component=kube-dns
                k8s-app=kube-dns
                pod-template-hash=377438151
Annotations:    scheduler.alpha.kubernetes.io/critical-pod: 
Status:         Running
IP:             10.55.142.166
Controlled By:  ReplicaSet/kube-dns-7cc87d595
Containers:
  kubedns:
    Container ID:  docker://d5332daee4b66f04f6ebe48198a539e9734abf36a1cf224f6d80911b7c8fec15
    Image:         602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-dns/kube-dns:1.14.10
    Image ID:      docker-pullable://602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-dns/kube-dns@sha256:7f2b2001f6571c4c44fc18005cee662254c3d931447a57cdbea65aaa4ae2dc16
    Ports:         10053/UDP, 10053/TCP, 10055/TCP
    Host Ports:    0/UDP, 0/TCP, 0/TCP
    Args:
      --domain=cluster.local.
      --dns-port=10053
      --config-dir=/kube-dns-config
      --v=2
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    255
      Started:      Wed, 21 Nov 2018 10:42:42 -0800
      Finished:     Wed, 21 Nov 2018 10:43:42 -0800
    Ready:          False
    Restart Count:  182
    Limits:
      memory:  170Mi
    Requests:
      cpu:      100m
      memory:   70Mi
    Liveness:   http-get http://:10054/healthcheck/kubedns delay=60s timeout=5s period=10s #success=1 #failure=5
    Readiness:  http-get http://:8081/readiness delay=3s timeout=5s period=10s #success=1 #failure=3
    Environment:
      PROMETHEUS_PORT:  10055
    Mounts:
      /kube-dns-config from kube-dns-config (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-dns-token-4x72t (ro)
  dnsmasq:
    Container ID:  docker://e76d3d6ad374fbd4bdd7f181616c8b6bc46499ac25bd0c7a7ecc357d0018a521
    Image:         602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-dns/dnsmasq-nanny:1.14.10
    Image ID:      docker-pullable://602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-dns/dnsmasq-nanny@sha256:acdfb44e37405dc9cfa25dca2a07ed97f35d0d77e34f18d8348475f9c181ad09
    Ports:         53/UDP, 53/TCP
    Host Ports:    0/UDP, 0/TCP
    Args:
      -v=2
      -logtostderr
      -configDir=/etc/k8s/dns/dnsmasq-nanny
      -restartDnsmasq=true
      --
      -k
      --cache-size=1000
      --no-negcache
      --log-facility=-
      --server=/cluster.local/127.0.0.1#10053
      --server=/in-addr.arpa/127.0.0.1#10053
      --server=/ip6.arpa/127.0.0.1#10053
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Wed, 21 Nov 2018 10:43:32 -0800
      Finished:     Wed, 21 Nov 2018 10:45:41 -0800
    Ready:          False
    Restart Count:  235
    Requests:
      cpu:        150m
      memory:     20Mi
    Liveness:     http-get http://:10054/healthcheck/dnsmasq delay=60s timeout=5s period=10s #success=1 #failure=5
    Environment:  <none>
    Mounts:
      /etc/k8s/dns/dnsmasq-nanny from kube-dns-config (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-dns-token-4x72t (ro)
  sidecar:
    Container ID:  docker://3bd43f5debaff66e9849fe78005ae87ad55741db4a3f809ca56ace21f9693a4d
    Image:         602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-dns/sidecar:1.14.10
    Image ID:      docker-pullable://602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/kube-dns/sidecar@sha256:5ff6aa8810bebe526ac20c752d7bd67c44f05fcb6f8759982fb782488a657110
    Port:          10054/TCP
    Host Port:     0/TCP
    Args:
      --v=2
      --logtostderr
      --probe=kubedns,127.0.0.1:10053,kubernetes.default.svc.cluster.local,5,SRV
      --probe=dnsmasq,127.0.0.1:53,kubernetes.default.svc.cluster.local,5,SRV
    State:          Running
      Started:      Wed, 21 Nov 2018 10:47:19 -0800
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Wed, 21 Nov 2018 10:45:42 -0800
      Finished:     Wed, 21 Nov 2018 10:47:18 -0800
    Ready:          True
    Restart Count:  255
    Requests:
      cpu:        10m
      memory:     20Mi
    Liveness:     http-get http://:10054/metrics delay=60s timeout=5s period=10s #success=1 #failure=5
    Environment:  <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-dns-token-4x72t (ro)
Conditions:
  Type           Status
  Initialized    True 
  Ready          False 
  PodScheduled   True 
Volumes:
  kube-dns-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      kube-dns
    Optional:  true
  kube-dns-token-4x72t:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  kube-dns-token-4x72t
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     CriticalAddonsOnly
                 node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                     From                                                  Message
  ----     ------     ----                    ----                                                  -------
  Warning  Unhealthy  38m (x911 over 18h)     kubelet, ip-10-55-142-134.us-west-2.compute.internal  Liveness probe failed: Get http://10.55.142.166:10054/metrics: dial tcp 10.55.142.166:10054: getsockopt: connection refused
  Warning  Unhealthy  28m (x1088 over 18h)    kubelet, ip-10-55-142-134.us-west-2.compute.internal  Readiness probe failed: Get http://10.55.142.166:8081/readiness: dial tcp 10.55.142.166:8081: getsockopt: connection refused
  Warning  Unhealthy  23m (x923 over 18h)     kubelet, ip-10-55-142-134.us-west-2.compute.internal  Liveness probe failed: Get http://10.55.142.166:10054/healthcheck/dnsmasq: dial tcp 10.55.142.166:10054: getsockopt: connection refused
  Normal   Killing    13m (x233 over 18h)     kubelet, ip-10-55-142-134.us-west-2.compute.internal  Killing container with id docker://dnsmasq:Container failed liveness probe.. Container will be killed and recreated.
  Warning  BackOff    8m47s (x3420 over 17h)  kubelet, ip-10-55-142-134.us-west-2.compute.internal  Back-off restarting failed container
  Warning  BackOff    3m51s (x4664 over 18h)  kubelet, ip-10-55-142-134.us-west-2.compute.internal  Back-off restarting failed container
```

We have also tried to deployed simpler example pods (for example, **httpd** with port **80**) but those fail with similar error as well. A sneak peak to kubelet reveals the following:

```
Nov 21 00:33:15 ip-10-55-142-134 systemd[1]: Starting Kubernetes Kubelet...
Nov 21 00:33:15 ip-10-55-142-134 systemd[1]: Started Kubernetes Kubelet.
Nov 21 00:33:27 ip-10-55-142-134 kubelet[2441]: Flag --allow-privileged has been deprecated, will be removed in a future version
Nov 21 00:33:27 ip-10-55-142-134 kubelet[2441]: I1121 00:33:27.772345    2441 feature_gate.go:226] feature gates: &{{} map[]}
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: Flag --allow-privileged has been deprecated, will be removed in a future version
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: I1121 00:33:28.118057    2441 feature_gate.go:226] feature gates: &{{} map[RotateKubeletServerCertificate:true]}
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: W1121 00:33:28.372992    2441 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: I1121 00:33:28.378658    2441 server.go:376] Version: v1.10.3
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: I1121 00:33:28.378714    2441 feature_gate.go:226] feature gates: &{{} map[RotateKubeletServerCertificate:true]}
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: I1121 00:33:28.473414    2441 aws.go:1026] Building AWS cloudprovider
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: I1121 00:33:28.473506    2441 aws.go:988] Zone not specified in configuration file; querying AWS metadata service
Nov 21 00:33:28 ip-10-55-142-134 kubelet[2441]: I1121 00:33:28.955293    2441 tags.go:76] AWS cloud filtering on ClusterID: dev
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.018723    2441 server.go:614] --cgroups-per-qos enabled, but --cgroup-root was not specified.  defaulting to /
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.019021    2441 container_manager_linux.go:242] container manager verified user specified cgroup-root exists: /
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.019036    2441 container_manager_linux.go:247] Creating Container Manager object based on Node Config: {RuntimeCgroupsName: SystemCgroupsName: KubeletCgroupsName: ContainerRuntime:docker CgroupsPerQOS:true CgroupRoot:/ 
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.019171    2441 container_manager_linux.go:266] Creating device plugin manager: true
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.019262    2441 state_mem.go:36] [cpumanager] initializing new in-memory state store
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.019500    2441 state_file.go:82] [cpumanager] state file: created new state file "/var/lib/kubelet/cpu_manager_state"
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.024360    2441 kubelet.go:298] Watching apiserver
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: W1121 00:33:29.308292    2441 kubelet_network.go:139] Hairpin mode set to "promiscuous-bridge" but kubenet is not enabled, falling back to "hairpin-veth"
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.308329    2441 kubelet.go:558] Hairpin mode set to "hairpin-veth"
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: W1121 00:33:29.308423    2441 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.308469    2441 client.go:75] Connecting to docker on unix:///var/run/docker.sock
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.308488    2441 client.go:104] Start docker client with request timeout=2m0s
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: W1121 00:33:29.311153    2441 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: W1121 00:33:29.313269    2441 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.313303    2441 docker_service.go:244] Docker cri networking managed by cni
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.331030    2441 docker_service.go:249] Docker Info: &{ID:TPEY:3ACM:ULSA:ILJC:R6ZG:ESYU:ZLJQ:FCCW:A5MA:P4NN:UMY7:VNNQ Containers:0 ContainersRunning:0 ContainersPaused:0 ContainersStopped:0 Images:0 Driver:overlay2 Driver
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.331839    2441 docker_service.go:262] Setting cgroupDriver to cgroupfs
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.366437    2441 remote_runtime.go:43] Connecting to runtime service unix:///var/run/dockershim.sock
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.475159    2441 kuberuntime_manager.go:186] Container runtime docker initialized, version: 18.09.0, apiVersion: 1.39.0
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: W1121 00:33:29.542957    2441 probe.go:215] Flexvolume plugin directory at /usr/libexec/kubernetes/kubelet-plugins/volume/exec/ does not exist. Recreating.
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.543383    2441 csi_plugin.go:61] kubernetes.io/csi: plugin initializing...
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.544345    2441 server.go:129] Starting to listen on 0.0.0.0:10250
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.545031    2441 server.go:299] Adding debug handlers to kubelet server.
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.559751    2441 fs_resource_analyzer.go:66] Starting FS ResourceAnalyzer
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.627803    2441 server.go:945] Started kubelet
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.628211    2441 status_manager.go:140] Starting to sync pod status with apiserver
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.628472    2441 kubelet.go:1782] Starting kubelet main sync loop.
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.628732    2441 kubelet.go:1799] skipping pod synchronization - [container runtime is down PLEG is not healthy: pleg was last seen active 2562047h47m16.854775807s ago; threshold is 3m0s]
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.628548    2441 volume_manager.go:247] Starting Kubelet Volume Manager
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.628567    2441 desired_state_of_world_populator.go:129] Desired state populator starts to run
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: E1121 00:33:29.629721    2441 kubelet.go:1282] Image garbage collection failed once. Stats initialization may not have completed yet: failed to get imageFs info: unable to find data for container /
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: W1121 00:33:29.632842    2441 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: E1121 00:33:29.634706    2441 kubelet.go:2130] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.735719    2441 kubelet_node_status.go:271] Setting node annotation to enable volume controller attach/detach
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.735756    2441 kubelet_node_status.go:327] Adding node label from cloud provider: beta.kubernetes.io/instance-type=m4.large
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.735767    2441 kubelet_node_status.go:338] Adding node label from cloud provider: failure-domain.beta.kubernetes.io/zone=us-west-2c
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.735775    2441 kubelet_node_status.go:342] Adding node label from cloud provider: failure-domain.beta.kubernetes.io/region=us-west-2
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.736154    2441 kubelet.go:1799] skipping pod synchronization - [container runtime is down]
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.750446    2441 kubelet_node_status.go:82] Attempting to register node ip-10-55-142-134.us-west-2.compute.internal
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.792546    2441 kubelet_node_status.go:271] Setting node annotation to enable volume controller attach/detach
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.792576    2441 kubelet_node_status.go:327] Adding node label from cloud provider: beta.kubernetes.io/instance-type=m4.large
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.792586    2441 kubelet_node_status.go:338] Adding node label from cloud provider: failure-domain.beta.kubernetes.io/zone=us-west-2c
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.792594    2441 kubelet_node_status.go:342] Adding node label from cloud provider: failure-domain.beta.kubernetes.io/region=us-west-2
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.797406    2441 cpu_manager.go:155] [cpumanager] starting with none policy
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.797424    2441 cpu_manager.go:156] [cpumanager] reconciling every 10s
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: I1121 00:33:29.797436    2441 policy_none.go:42] [cpumanager] none policy: Start
```
I do notice these lines:
```
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: E1121 00:33:29.629721    2441 kubelet.go:1282] Image garbage collection failed once. Stats initialization may not have completed yet: failed to get imageFs info: unable to find data for container /
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: W1121 00:33:29.632842    2441 cni.go:171] Unable to update cni config: No networks found in /etc/cni/net.d
Nov 21 00:33:29 ip-10-55-142-134 kubelet[2441]: E1121 00:33:29.634706    2441 kubelet.go:2130] Container runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized
```
Not sure if it is signifcant?

Also, running /opt/cni/bin/aws-cni-support.sh produces:
```
/opt/cni/bin/aws-cni-support.sh 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   936  100   936    0     0   184k      0 --:--:-- --:--:-- --:--:--  228k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   145  100   145    0     0    98k      0 --:--:-- --:--:-- --:--:--  141k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   106  100   106    0     0  75391      0 --:--:-- --:--:-- --:--:--  103k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    83  100    83    0     0  15770      0 --:--:-- --:--:-- --:--:-- 16600
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    28  100    28    0     0  18791      0 --:--:-- --:--:-- --:--:-- 28000
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6130  100  6130    0     0   979k      0 --:--:-- --:--:-- --:--:--  997k
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed to connect to localhost port 10255: Connection refused
```
This means `curl http://localhost:10255/pods  > ${LOG_DIR}/kubelet.output` fails because there is no listening process on port 10255.
After adjusting `/opt/cni/bin/aws-cni-support.sh` to fit Ubuntu16.04, we have following output:
[aws-cni-support.tar.gz](https://github.com/aws/amazon-vpc-cni-k8s/files/2605478/aws-cni-support.tar.gz)

We suspect this may be a routing issue with Docker on the worker node but can't be so sure. Any suggestion is highly appreciated.

**Comment by nickdgriffin** [comment]
How are your workers setup? Is it the template from [here](https://github.com/awslabs/amazon-eks-ami/blob/master/amazon-eks-nodegroup.yaml) or something custom?

**Comment by huynt1979** [comment]
@nickdgriffin Yes. I have set up both EKS and worker nodes according to the guide. We use Terraform so the CloudFormation file is not used directly. But I'm confident that all the required IAMs, Security Groups... are strictly followed. K8s RBAC for worker nodes specified [aws-auth-cm](https://amazon-eks.s3-us-west-2.amazonaws.com/cloudformation/2018-11-07/aws-auth-cm.yaml) is applied. Packer image and bootstrap run fine without issues. I don't see any security related errors in kubelet log. 

Let me know if you need more info.

**Comment by huynt1979** [comment]
I've managed to get aws-cni working with our EKS cluster. There are really 2 changes that I did:
- Use kubelet & kubectl from https://storage.googleapis.com/kubernetes-release/release/${RELEASE} with RELEASE matches my EKS version - **1.10.3**. Before these binaries were downloaded from **amazon-eks** bucket.

- Use the default pause-container image that comes with K8s. Before, it was referring to 602401143452.dkr.ecr.$AWS_DEFAULT_REGION.amazonaws.com/eks/pause-amd64:3.1

This led me to suspect the followings:
- The kubelet binary in amazon-eks S3 might not work.
- The eks pause-container has issue.

Either way, it works for me now. And as my setup uses the standard binaries and images coming from Google, it's actually preferrably in my case.

Hope this help. Close issue.
